static/faq/lenguaje-c.md

```diff
@@ -182,3 +182,148 @@ typedef void funcionvoid_t(void*);
+
+Es importante destacar que los `enum` sirven para hacer más claro y legible el código y son una herramienta **para el programador**. Como tales no hay herramientas para convertir, por ejemplo, la etiqueta `ANIMALIA` en la cadena `"ANIMALIA"` ni hay herramientas para leer un enumerativo.
+
+### Diccionarios para traducir enumerativos a cadenas (o a cualquier tipo)
+
+Si necesitáramos traducir entre la representación interna del enumerativo y otras representaciones adecuadas para interactuar con el exterior vamos a aprovechar la propiedad de la numeración de las etiquetas. Si un enumerativo tiene `N` etiquetas sabemos que las mismas serán numéricas y numeradas entre `0` y `N-1` por lo que se hace inmediato aparear enumerativos a arreglos. Es decir, se pueden construir "diccionarios" que _asocian_ un valor arbitrario en algún tipo con un enumerativo según la posición en el mismo del valor.
+
+Por ejemplo:
+
+```cpp
+const char *nombres_reino[] = {"animalia", "plantae", "fungi", "bacteria", "protista", "archaea"};
+
+char *reino_a_cadena(enum reino r) {
+    return nombres_reino[r];
+}
+```
+
+Notar que la precondición implícita de la función es asumir que se cumple el invariante del enumerativo, es decir, que el mismo contiene una etiqueta válida de la definición del `enum`.
+
+Si quisiéramos realizar el proceso contrario, esto es, implementar una función `enum reino cadena_a_reino(const char *s);` podríamos iterar comparando los elementos de `nombre_reino` hasta que encontremos la cadena y sencillamente devolver el índice del vector (que es entero y en el rango del enumerativo, por lo que no hay inconveniente en castearlo a tipo `enum reino`). Ahora bien, van a surgir dos complicaciones: La primera es que no tenemos manera de preguntarle al enumerativo cuántas etiquetas posee (esto podría saltearse preguntándoselo al vector con `sizeof(nombres_reino)/sizeof(nombres_reino[0])`, la segunda es que no sabríamos qué devolver en el caso de que la cadena `s` no coincidiera con ningún elemento de `nombres_reino`. Si bien ambos problemas pueden evadirse probablemente convenga manejarlos de otra forma.
+
+### Diccionarios para traducir a enumerativo
+
+En el caso de que supiéramos que lo que hay que traducir viene bien formado y ya pasó por un proceso previo de verificación que garantiza la correctitud de lo datos (suele ser una etapa recomendable en cualquier programa) bien podemos hacer el reverso que para traducir de enumerativo a otra cosa:
+
+```cpp
+// Precondición: La cadena pertenece a nombre_reino
+enum reino cadena_a_reino(const char *s) {
+    size_t cantidad = sizeof(nombre_reino)/sizeof(nombre_reino[0]);
+    for(size_t i = 0; i < cantidad; i++)
+        if(!strcmp(nombre_reino[i], s))
+            return i;
```

Comment:
Esto no es un diccionario. (Y los de abajo son diccionarios divididos en dos arreglos. Me resulta muy confuso que estén cuando luego _sí_ se introducen los arreglos de structs.)

Pero bueno: en realidad le tengo un poco de alergia sobre asumir que nadie hizo `= 1` en la declaración del enum.

Hay muchos, muchos ejemplos de código en esta FAQ. Si es código que va a terminar en TPs, preferiría una síntesis que sea poco liosa. Tal como está ahora, es un tratado de todas las maneras en que se puede abordar el problema. 
Estoy un poco embotado y no puedo seguir la review bien, pero... ¿qué te parece algo del estilo...

```cpp
// Siempre *dos* arreglos: el canónico cadena → enum, y el interno enum →todos los datos:
static const ... cmd_names[] = {{"name1", CMD1}, ... };
static const ... cmd_data[] = {
    [CMD1] = {...},
    [CMD2] = {...},
};
 
bool parse_comando(const char *s, enum cmd *ret);
```

Lo que odio son los structs (not shown), pero me parece que les vamos a matar si los hacemos anónimos.